crates/pop-common/src/api.rs

```diff
@@ -0,0 +1,123 @@
+// SPDX-License-Identifier: GPL-3.0
+
+use crate::APP_USER_AGENT;
+use reqwest::{IntoUrl, Response};
+use std::{
+	error::Error as _,
+	ops::Deref,
+	sync::{Arc, Mutex},
+	time::{SystemTime, SystemTimeError},
+};
+use thiserror::Error;
+use tokio::{
+	sync::{AcquireError, Semaphore},
+	time::{Duration, Instant},
+};
+
+pub(crate) struct ApiClient {
+	permits: Arc<Semaphore>,
+	token: Option<String>,
+	rate_limits: Arc<Mutex<RateLimits>>,
+}
+impl ApiClient {
+	pub(crate) fn new(max_concurrent: usize, token: Option<String>) -> Self {
+		Self {
+			permits: Arc::new(Semaphore::new(max_concurrent)),
+			token,
+			rate_limits: Arc::new(Mutex::new(RateLimits::default())),
+		}
+	}
+
+	pub(crate) async fn get<U: IntoUrl>(&self, url: U) -> Result<Response, Error> {
+		let _permit = self.permits.acquire().await?;
```

Comment:
https://docs.github.com/en/rest/using-the-rest-api/best-practices-for-using-the-rest-api